File: backend/revert.py
import os
import re

entity_dir = r"e:\ManiStayBite\backend\src\main\java\com\example\staybite_backend\entity"
repo_dir = r"e:\ManiStayBite\backend\src\main\java\com\example\staybite_backend\repository"
controller_dir = r"e:\ManiStayBite\backend\src\main\java\com\example\staybite_backend\controller"
service_dir = r"e:\ManiStayBite\backend\src\main\java\com\example\staybite_backend\service"

def replace_in_file(filepath, replacements):
    with open(filepath, 'r') as f:
        content = f.read()
    
    new_content = content
    for old, new in replacements:
        if callable(old):
            new_content = old(new_content)
        else:
            new_content = new_content.replace(old, new)
            
    if new_content != content:
        with open(filepath, 'w') as f:
            f.write(new_content)
        print(f"Updated {filepath}")

# 1. Revert Entities
skip_entities = ["AuditLog.java", "Role.java"]
for filename in os.listdir(entity_dir):
    if filename.endswith(".java") and filename not in skip_entities:
        filepath = os.path.join(entity_dir, filename)
        table_name = filename.replace(".java", "").lower() + "s"
        if filename == "Hotel.java": table_name = "hotels"
        elif filename == "User.java": table_name = "users"
        elif filename == "Room.java": table_name = "rooms"
        elif filename == "Order.java": table_name = "orders"
        elif filename == "Attendance.java": table_name = "attendance"
        elif filename == "EventBooking.java": table_name = "event_bookings"
        elif filename == "Food.java": table_name = "foods"
        elif filename == "ResortService.java": table_name = "resort_services"
        elif filename == "Review.java": table_name = "reviews"
        elif filename == "Subscription.java": table_name = "subscriptions"
        elif filename == "SupportTicket.java": table_name = "support_tickets"
        elif filename == "Wallet.java": table_name = "wallets"

        replacements = [
            ("import org.springframework.data.annotation.Id;\nimport org.springframework.data.mongodb.core.mapping.Document;\nimport org.springframework.data.mongodb.core.mapping.DBRef;", "import jakarta.persistence.*;"),
            ("import org.springframework.data.annotation.Id;\nimport org.springframework.data.mongodb.core.mapping.Document;", "import jakarta.persistence.*;"),
            (f'@Document(collection = "{table_name}")', f'@Entity\n@Table(name = "{table_name}")'),
            ("@Id\n    private String ", "@Id\n    @GeneratedValue(strategy = GenerationType.IDENTITY)\n    private Long "),
            ("@DBRef", "@ManyToOne"), # Default, will manually fix ones that were OneToOne or needed JoinColumn
            ("private String orderId;", "private Long orderId;"),
            ("private String roomId;", "private Long roomId;"),
            ("private String reviewId;", "private Long reviewId;"),
            ("private String foodId;", "private Long foodId;"),
            ("private String eventId;", "private Long eventId;"),
            ("private String entityId;", "private Long entityId;"),
        ]
        replace_in_file(filepath, replacements)

# 2. Revert Repositories
skip_repos = ["AuditLogRepository.java"]
for filename in os.listdir(repo_dir):
    if filename.endswith(".java") and filename not in skip_repos:
        filepath = os.path.join(repo_dir, filename)
        replacements = [
            ("org.springframework.data.mongodb.repository.MongoRepository", "org.springframework.data.jpa.repository.JpaRepository"),
            ("extends MongoRepository<", "extends JpaRepository<"),
            (", String>", ", Long>"),
            ("String employeeId", "Long employeeId"),
            ("String userId", "Long userId"),
            ("String hotelId", "Long hotelId"),
            ("String entityId", "Long entityId"),
        ]
        replace_in_file(filepath, replacements)

# 3. Revert Controllers and Services
for directory in [controller_dir, service_dir]:
    for filename in os.listdir(directory):
        if filename.endswith(".java"):
            filepath = os.path.join(directory, filename)
            replacements = [
                ("String userId", "Long userId"),
                ("String id", "Long id"),
                ("String orderId", "Long orderId"),
                ("String ticketId", "Long ticketId"),
                ("String hotelId", "Long hotelId"),
                ("String employeeId", "Long employeeId"),
                ("String attendanceId", "Long attendanceId"),
            ]
            replace_in_file(filepath, replacements)

print("Revert completed.")

File: backend/Revert.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

public class Revert {
    public static void main(String[] args) throws Exception {
        String entityDir = "src/main/java/com/example/staybite_backend/entity";
        String repoDir = "src/main/java/com/example/staybite_backend/repository";
        String controllerDir = "src/main/java/com/example/staybite_backend/controller";
        String serviceDir = "src/main/java/com/example/staybite_backend/service";

        System.out.println("Starting Revert...");

        // 1. Entities
        File eDir = new File(entityDir);
        if (eDir.exists()) {
            for (File f : eDir.listFiles()) {
                String name = f.getName();
                if (name.endsWith(".java") && !name.equals("AuditLog.java") && !name.equals("Role.java")) {
                    String content = new String(Files.readAllBytes(f.toPath()));
                    
                    content = content.replace("import org.springframework.data.annotation.Id;\nimport org.springframework.data.mongodb.core.mapping.Document;\nimport org.springframework.data.mongodb.core.mapping.DBRef;", "import jakarta.persistence.*;");
                    content = content.replace("import org.springframework.data.annotation.Id;\nimport org.springframework.data.mongodb.core.mapping.Document;", "import jakarta.persistence.*;");
                    
                    String tableName = name.replace(".java", "").toLowerCase() + "s";
                    if (name.equals("Hotel.java")) tableName = "hotels";
                    else if (name.equals("User.java")) tableName = "users";
                    else if (name.equals("Room.java")) tableName = "rooms";
                    else if (name.equals("Order.java")) tableName = "orders";
                    else if (name.equals("Attendance.java")) tableName = "attendance";
                    else if (name.equals("EventBooking.java")) tableName = "event_bookings";
                    else if (name.equals("Food.java")) tableName = "foods";
                    else if (name.equals("ResortService.java")) tableName = "resort_services";
                    else if (name.equals("Review.java")) tableName = "reviews";
                    else if (name.equals("Subscription.java")) tableName = "subscriptions";
                    else if (name.equals("SupportTicket.java")) tableName = "support_tickets";
                    else if (name.equals("Wallet.java")) tableName = "wallets";
                    
                    content = content.replace("@Document(collection = \"" + tableName + "\")", "@Entity\n@Table(name = \"" + tableName + "\")");
                    content = content.replace("@Id\n    private String ", "@Id\n    @GeneratedValue(strategy = GenerationType.IDENTITY)\n    private Long ");
                    content = content.replace("private String orderId;", "private Long orderId;");
                    content = content.replace("private String roomId;", "private Long roomId;");
                    content = content.replace("private String reviewId;", "private Long reviewId;");
                    content = content.replace("private String foodId;", "private Long foodId;");
                    content = content.replace("private String eventId;", "private Long eventId;");
                    
                    // Specific JoinColumn fixes
                    if (name.equals("User.java")) {
                        content = content.replace("@DBRef\n    private Hotel hotel;", "@ManyToOne\n    @JoinColumn(name = \"hotel_id\")\n    private Hotel hotel;");
                    } else if (name.equals("Hotel.java")) {
                        content = content.replace("@DBRef\n    private User owner;", "@ManyToOne\n    @JoinColumn(name = \"owner_id\", nullable = false)\n    private User owner;");
                    } else if (name.equals("Room.java") || name.equals("Food.java") || name.equals("Order.java")) {
                        content = content.replace("@DBRef\n    private Hotel hotel;", "@ManyToOne\n    @JoinColumn(name = \"hotel_id\")\n    private Hotel hotel;");
                        if (name.equals("Order.java")) {
                            content = content.replace("@DBRef\n    private User user;", "@ManyToOne\n    @JoinColumn(name = \"user_id\", nullable = false)\n    private User user;");
                            content = content.replace("@DBRef\n    private User employee;", "@ManyToOne\n    @JoinColumn(name = \"employee_id\")\n    private User employee;");
                        }
                    } else if (name.equals("Attendance.java")) {
                        content = content.replace("@DBRef\n    private User employee;", "@ManyToOne\n    @JoinColumn(name = \"employee_id\", nullable = false)\n    private User employee;");
                    } else if (name.equals("EventBooking.java") || name.equals("Review.java") || name.equals("SupportTicket.java") || name.equals("ResortService.java")) {
                        content = content.replace("@DBRef\n    private User user;", "@ManyToOne\n    @JoinColumn(name = \"user_id\", nullable = false)\n    private User user;");
                        if(name.equals("ResortService.java")) {
                            content = content.replace("@JoinColumn(name = \"user_id\", nullable = false)", "@JoinColumn(name = \"user_id\")");
                        }
                    } else if (name.equals("Subscription.java") || name.equals("Wallet.java")) {
                        if(name.equals("Subscription.java")) {
                            content = content.replace("@DBRef\n    private Hotel hotel;", "@OneToOne\n    @JoinColumn(name = \"hotel_id\", nullable = false)\n    private Hotel hotel;");
                        } else {
                            content = content.replace("@DBRef\n    private User user;", "@OneToOne\n    @JoinColumn(name = \"user_id\", nullable = false)\n    private User user;");
                        }
                    }
                    
                    if (name.equals("Review.java")) {
                        content = content.replace("private String entityId;", "private Long entityId;");
                    }

                    Files.write(f.toPath(), content.getBytes());
                    System.out.println("Updated " + f.getName());
                }
            }
        }

        // 2. Repositories
        File rDir = new File(repoDir);
        if (rDir.exists()) {
            for (File f : rDir.listFiles()) {
                String name = f.getName();
                if (name.endsWith(".java") && !name.equals("AuditLogRepository.java")) {
                    String content = new String(Files.readAllBytes(f.toPath()));
                    content = content.replace("org.springframework.data.mongodb.repository.MongoRepository", "org.springframework.data.jpa.repository.JpaRepository");
                    content = content.replace("extends MongoRepository<", "extends JpaRepository<");
                    content = content.replace(", String>", ", Long>");
                    content = content.replace("String employeeId", "Long employeeId");
                    content = content.replace("String userId", "Long userId");
                    content = content.replace("String hotelId", "Long hotelId");
                    content = content.replace("String entityId", "Long entityId");
                    Files.write(f.toPath(), content.getBytes());
                    System.out.println("Updated " + f.getName());
                }
            }
        }

        // 3. Controllers and Services
        String[] dirs = {controllerDir, serviceDir};
        for (String d : dirs) {
            File dir = new File(d);
            if (dir.exists()) {
                for (File f : dir.listFiles()) {
                    if (f.getName().endsWith(".java")) {
                        String content = new String(Files.readAllBytes(f.toPath()));
                        content = content.replace("String userId", "Long userId");
                        content = content.replace("String id", "Long id");
                        content = content.replace("String orderId", "Long orderId");
                        content = content.replace("String ticketId", "Long ticketId");
                        content = content.replace("String hotelId", "Long hotelId");
                        content = content.replace("String employeeId", "Long employeeId");
                        content = content.replace("String attendanceId", "Long attendanceId");
                        Files.write(f.toPath(), content.getBytes());
                        System.out.println("Updated " + f.getName());
                    }
                }
            }
        }

        System.out.println("Done.");
    }
}
